README.md
## geoip -- GeoIP lookup service.

## Table of Contents
- [Installation](#installation)
  - [Ubuntu/Debian](#ubuntudebian)
  - [CentOS/Redhat](#centosredhat)
  - [Manual Install](#manual-install)
  - [Build from source](#build-from-source)
- [Usage](#usage)
  - [Example](#example)
- [Contributing](#contributing)
- [License](#license)

## Installation

Check out the [releases](https://github.com/lrstanley/geoip/releases)
page for prebuilt versions. Below are example commands of how you would install
the utility. Some of the more popular OS/distro steps are provided below, but
there are more released versions on the releases page previously mentioned.

### Ubuntu/Debian

```console
$ wget https://liam.sh/ghr/geoip_0.1.3_linux_amd64.deb
$ dpkg -i geoip_0.1.3_linux_amd64.deb
```

### CentOS/Redhat

```console
$ yum localinstall https://liam.sh/ghr/geoip_0.1.3_linux_amd64.rpm
```

### Manual Install

```console
$ wget https://liam.sh/ghr/geoip_0.1.3_linux_amd64.tar.gz
$ tar -C /usr/bin/ -xzvf geoip_0.1.3_linux_amd64.tar.gz geoip
$ chmod +x /usr/bin/geoip
```

### Build From Source

Dependencies (to build from source only):

   * [Go](https://golang.org/doc/install) (1.9 or greater, though latest
   preferred). Ensure your `$GOPATH` is setup.
   * [NodeJS](https://nodejs.org/en/download/) (v6 or greater)

Setup:

```console
$ go get -d -u github.com/lrstanley/geoip
$ cd $GOPATH/src/github.com/lrstanley/geoip
# this will show you all of the available options (to fetch dependencies,
# run in debug mode, etc.)
$ make help
$ make
$ ./geoip --help
```

For active development:

```console
$ make fetch # make sure this is ran at least once to fetch all dependencies.
$ make debug
# run this in a different window. this will rebundle the frontend assets on
# change.
$ make generate-watch
```

## Usage

```console
$ geoip --help
Usage:
  geoip [OPTIONS]

Application Options:
  -d, --debug          enable exception display and pprof endpoints (warn: dangerous)
  -q, --quiet          disable verbose output
      --db=            path to read/store Maxmind DB (default: geoip.db)
      --interval=      interval of time between database update checks (default: 2h)
      --update-url=    maxmind database file download location (must be gzipped) (default: http://geolite.maxmind.com/download/geoip/database/GeoLite2-City.mmdb.gz)
  -v, --version        print the version and compilation date

Cache Options:
      --cache.size=    total number of lookups to keep in ARC cache (50% most recent, 50% most requested) (default: 500)
      --cache.expire=  expiration time of cache (default: 20m)

HTTP Options:
  -b, --http.bind=     address and port to bind to (default: :8080)
      --http.proxy     obey X-Forwarded-For headers (warn: dangerous, make sure to only bind to localhost)
      --http.throttle= limit total max concurrent requests across all connections
      --http.limit=    number of requests/ip/hour (default: 2000)
      --http.cors=     cors origin domain to allow with https?:// prefix (empty => '*'; use flag multiple times)

TLS Options:
      --http.tls.use   enable tls
      --http.tls.cert= path to ssl certificate
      --http.tls.key=  path to ssl key

DNS Lookup Options:
      --dns.timeout=   max allowed duration when looking up hostnames (may cause queries to be slow) (default: 2s)
      --dns.resolver=  resolver (in host:port form) to use for dns lookups (doesn't work with windows and plan9) (can be used multiple times)
      --dns.uselocal   adds local (system) resolvers to the list of resolvers to use

Help Options:
  -h, --help           Show this help message

```

### Example

```console
$ geoip --cache.size 1000 --http.bind "localhost:8080" --http.proxy --http.limit 15000 --dns.resolver 8.8.8.8 --dns.resolver 8.8.4.4
```

## Contributing

Please review the [CONTRIBUTING](CONTRIBUTING.md) doc for submitting issues/a guide
on submitting pull requests and helping out.

## License

    LICENSE: The MIT License (MIT)
    Copyright (c) 2015 Liam Stanley <me@liamstanley.io>

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in
    all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]

## [0.1.3] - 2018-04-11
### Changed
- fix: `/api/self` queries not properly working when behind an http proxy.

## [0.1.2] - 2018-04-10
### Changed
- Builds should now be fully statically linked.
- jQuery to jQuery-slim (smaller download size, faster).
- SemanticUI **2.2.13** to **2.3.1**.
- LeafletJS **1.2.0** to **1.3.1**.

## [0.1.1] - 2018-01-17
### Added
- "Source on Github" sidebar link.
- `--http.throttle` flag, allowing the limitation of concurrent requests.

### Changed
- fix: `Content-Type` header for filtered requests was improperly being set
to `application/json` (should have been `text/plain`.)
- fix: spinner icon for each entry on the bulk lookup page was misaligned.

## [0.1.0] - 2017-10-13
- Initial release.

[Unreleased]: https://github.com/lrstanley/geoip/compare/v0.1.3...HEAD
[0.1.3]: https://github.com/lrstanley/geoip/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/lrstanley/geoip/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/lrstanley/geoip/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/lrstanley/geoip/tree/v0.1.0


main.go
// Copyright (c) Liam Stanley <me@liamstanley.io>. All rights reserved. Use
// of this source code is governed by the MIT license that can be found in
// the LICENSE file.

package main

import (
	"context"
	"fmt"
	"io/ioutil"
	"log"
	"math/rand"
	"net"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	"github.com/bluele/gcache"
	gflags "github.com/jessevdk/go-flags"
)

// Should be automatically added by goreleaser.
var (
	version = "unknown"
	commit  = "unknown"
	date    = "unknown"
)

type Flags struct {
	Debug          bool          `short:"d" long:"debug" description:"enable exception display and pprof endpoints (warn: dangerous)"`
	Quiet          bool          `short:"q" long:"quiet" description:"disable verbose output"`
	DBPath         string        `long:"db" description:"path to read/store Maxmind DB" default:"geoip.db"`
	UpdateInterval time.Duration `long:"interval" description:"interval of time between database update checks" default:"2h"`
	UpdateURL      string        `long:"update-url" description:"maxmind database file download location (must be gzipped)" default:"http://geolite.maxmind.com/download/geoip/database/GeoLite2-City.mmdb.gz"`
	Cache          struct {
		Size   int           `long:"size" description:"total number of lookups to keep in ARC cache (50% most recent, 50% most requested)" default:"500"`
		Expire time.Duration `long:"expire" description:"expiration time of cache" default:"20m"`
	} `group:"Cache Options" namespace:"cache"`
	HTTP struct {
		Bind     string   `short:"b" long:"bind" description:"address and port to bind to" default:":8080"`
		Proxy    bool     `long:"proxy" description:"obey X-Forwarded-For headers (warn: dangerous, make sure to only bind to localhost)"`
		Throttle int      `long:"throttle" description:"limit total max concurrent requests across all connections"`
		Limit    int      `long:"limit" description:"number of requests/ip/hour" default:"2000"`
		CORS     []string `long:"cors" description:"cors origin domain to allow with https?:// prefix (empty => '*'; use flag multiple times)"`
		TLS      struct {
			Use  bool   `long:"use" description:"enable tls"`
			Cert string `long:"cert" description:"path to ssl certificate"`
			Key  string `long:"key" description:"path to ssl key"`
		} `group:"TLS Options" namespace:"tls"`
	} `group:"HTTP Options" namespace:"http"`
	DNS struct {
		Timeout   time.Duration `long:"timeout" description:"max allowed duration when looking up hostnames (may cause queries to be slow)" default:"2s"`
		Resolvers []string      `long:"resolver" description:"resolver (in host:port form) to use for dns lookups (doesn't work with windows and plan9) (can be used multiple times)"`
		Local     bool          `long:"uselocal" description:"adds local (system) resolvers to the list of resolvers to use"`
	} `group:"DNS Lookup Options" namespace:"dns"`
	Version bool `short:"v" long:"version" description:"print the version and compilation date"`
}

var (
	flags    Flags
	logger   = log.New(ioutil.Discard, "", log.LstdFlags|log.Lshortfile)
	db       *DB
	arc      gcache.Cache
	resolver *net.Resolver
)

func main() {
	parser := gflags.NewParser(&flags, gflags.HelpFlag)
	_, err := parser.Parse()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	if flags.Version {
		fmt.Printf("geoip version %q (compiled: %q, commit: %q)\n", version, date, commit)
		os.Exit(0)
	}

	if !flags.Quiet {
		logger.SetOutput(os.Stdout)
	}

	db = &DB{path: flags.DBPath}
	arc = gcache.New(flags.Cache.Size).ARC().Expiration(flags.Cache.Expire).Build()

	if len(flags.DNS.Resolvers) == 0 {
		resolver = net.DefaultResolver
	} else {
		resolver = &net.Resolver{PreferGo: true, Dial: customResolver}
	}

	go func() {
		var needsUpdate bool
		var err error
		for {
			logger.Println("checking for database updates")
			needsUpdate, err = db.checkForUpdates()
			if needsUpdate {
				if err != nil {
					logger.Printf("database needs update due to error (%s)", err)
				} else {
					logger.Println("database needs update")
				}

				if err = db.update(flags.UpdateURL); err != nil {
					logger.Println(err)
				}
			} else {
				logger.Println("no database updates needed")
			}

			time.Sleep(flags.UpdateInterval)
		}
	}()

	httpCloser := make(chan struct{})
	go initHTTP(httpCloser)

	catch()
	close(httpCloser)
	fmt.Println("exiting")
}

func catch() {
	signals := make(chan os.Signal, 1)
	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM, os.Interrupt)

	fmt.Println("listening for signal. CTRL+C to quit.")
	<-signals
	fmt.Println("\ninvoked termination, cleaning up")
}

func customResolver(ctx context.Context, network, address string) (net.Conn, error) {
	var index int

	if flags.DNS.Local {
		index = rand.Intn(len(flags.DNS.Resolvers) + 1)
	} else {
		// Generate a random number, which is used to select a resolver.
		// However, if the number generated is out of the bounds of the
		// amount of resolvers, use the system resolver, since they
		// requested it.
		index = rand.Intn(len(flags.DNS.Resolvers))
	}

	if index == len(flags.DNS.Resolvers) {
		return net.Dial(network, address)
	}

	addr := flags.DNS.Resolvers[index]

	if strings.Contains(addr, ":") {
		return net.Dial(network, addr)
	}
	return net.Dial(network, addr+":53")
}
